command: spack config blame

tgamblin opened the pull request:
Have you ever wondered why Spack is behaving a certain way or using a particular external package?  Do you wonder what [configuration scope](https://spack.readthedocs.io/en/latest/configuration.html) a particular setting came from?  Now you can find out! `spack config blame <config file>`  will print out the YAML for a Spack configuration section, annotated with the file and line where each setting came from.  So, now you can track down exactly why your configuration works a certain way.

Here's an example:

<img width="1275" alt="screen shot 2018-05-09 at 10 25 26 pm" src="https://user-images.githubusercontent.com/299842/39853662-61ca68d2-53d8-11e8-8058-c0287a00eac3.png">

In this picture, you can the full, merged `config.yaml` on the right.  On the left, you can see that some settings come from hard-coded builtings (`_builtin`), some come from the default Spack `config.yaml`, some come from `~/.spack/config.yaml`, and some come from the command line.

This works for any config file, so you can do the same thing with `spack config blame packages` or `spack config blame compilers`, etc.

tgamblin commented:
@alalazo @scheibelp @adamjstewart: This is ready to go now that #7774 is merged.

scheibelp commented on lib/spack/spack/util/spack_yaml.py:
```diff
@@ -60,10 +65,56 @@ class syaml_str(str):
+    __repr__ = str.__repr__
+
+
+#: primitive type -> syaml type mapping
```
I think this should be reversed: `syaml_type -> primitive type mapping`

scheibelp replied:
This is a map *from* syaml type *to* primitive type. The comment is confusing as is (or did you disagree with that?)

scheibelp commented on lib/spack/spack/util/spack_yaml.py:
```diff
@@ -204,6 +256,69 @@ def ignore_aliases(self, _data):
+
+
+class FirstPassDumper(OrderedLineDumper):
+    """This is a first pass for AnnotatedDumper.
```
I think I'm missing something: I don't see an `AnnotatedDumper` class declared anywhere

scheibelp commented on lib/spack/spack/util/spack_yaml.py:
```diff
@@ -204,6 +256,69 @@ def ignore_aliases(self, _data):
+
+#: Global for interactions between FirstPassDumper and AnnotatedDumper
+#: This is nasty but YAML doesn't give us many ways to pass arguments.
+_annotations = []
```
The only place I see `_annotations` being written is by `FirstPassDumper`, the only place I see it read is in `dump_annotations`, which also declares `FirstPassDumper`. So I'm not sure why this has to be global.

tgamblin commented:
@scheibelp: this is updated and ready for re-review

scheibelp commented on lib/spack/spack/test/util/spack_yaml.py:
```diff
@@ -0,0 +1,70 @@
+
+
+def check_blame(element, scope=None):
+    """Get an element from a config file and check its blame information.
```
IMO the phrasing

> a config file

is somewhat confusing since this specifically only tests the contents of the `config` section config file. The explanation itself may need some extra detail (in part because it is confusing for one of the sections to be called `config`, although that's not an issue with this PR specifically):

> Test annotations for the `config` section of the configuration.

scheibelp commented on lib/spack/spack/test/util/spack_yaml.py:
```diff
@@ -0,0 +1,70 @@
+    element = spack.config.get('config:%s' % element)
+
+    if scope is None:
+        annotation = '%s:%d' % (element._start_mark.name,
```
IMO this is somewhat roundabout since the test is using the output generated by the test to verify the output (e.g. constructing the annotation by accessing the parsed element). I imagine it would be fragile to make an assertion about a specific line number but could a regex be generated to verify it:

`"{0}:\d\s+{1}".format(expected_section_name, cfg_var_name)`

(and you can just use parameters to the check function vs. extracting from the element)

Also: what line number *is* this supposed to be in `test_config_blame_with_override`? I thought the `override` scope doesn't have an associated file; if I'm right about that, would it make sense to test this in one case where there is an associated file?

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -608,6 +613,26 @@ def _mark_overrides(data):
+def _mark_internal(data, name):
+    """Add a simple name mark to raw YAML/JSON data.
+
+    This is used by `spack config blame` to show where config lins came from.
```
typo: lines

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -639,13 +664,28 @@ def they_are(t):
+        # ensures we can get the actual source key objects from dest keys
+        for dk in dest.keys():
+            if dk in key_marks:
+                syaml.mark(dk, key_marks[dk])
```
Super minor point (because this is already a huge a leap forward what we have right now): a key of a dictionary that is effectively a merge of more files is attributed to one of them. Example:
```console
$ spack config blame packages
---                                                                     packages:
/home/mculpo/PycharmProjects/spack/etc/spack/packages.yaml:2              all:
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      compiler:
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - gcc
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - intel
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - pgi
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - clang
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - xl
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - nag
/home/mculpo/PycharmProjects/spack/etc/spack/packages.yaml:3                providers:
<omitted>
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:25        gl:
/home/mculpo/PycharmProjects/spack/etc/spack/packages.yaml:5                  - mesa
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:26        glu:
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:26        - mesa-glu
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:26        - openglu
<omitted>
```
Above I think that `all` and `providers` should either have no tag, or have a tag that says they are the result of a merge.

tgamblin replied:
Yeah, this is a reasonable point.  I guess it's somewhat obvious to me that it's a merge if the values have multi-file annotations.

As I don't have a quick fix for this, I think I will leave it for a future PR.

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -608,6 +613,26 @@ def _mark_overrides(data):
+
+    if syaml.markable(d):
+        d._start_mark = yaml.Mark(name, None, None, None, None, None)
+        d._end_mark = yaml.Mark(name, None, None, None, None, None)
```
Maybe a keyword binding here would be more readable:
```python
d._end_mark = yaml.Mark(name=name, index=None, line=None, column=None, buffer=None, pointer=None)
```
?

tgamblin replied:
I'm trying to stick with the way the original YAML code works.  While I agree that it could be more readable, I figure that if they don't rely on names of parameters, we shouldn't either.  The YAML code is pretty nasty.

alalazo commented on lib/spack/spack/util/spack_yaml.py:
```diff
@@ -60,10 +65,56 @@ class syaml_str(str):
+    if hasattr(node, 'start_mark'):
+        obj._start_mark = node.start_mark
+    elif hasattr(node, '_start_mark'):
+        obj._start_mark = node._start_mark
```
Side note on style (not directly introduced by this PR). We are doing a lot of reads and writes of two "private" attributes:
```python
obj._start_mark
obj._end_mark
``` 
Probably it would be nicer to make the attribute names not start with an underscore.

tgamblin replied:
This was done to avoid conflicts with YAML's own markings.  I suppose I could've picked another prefix, but it starts with `_` to avoid conflicts with vanilla YAML, not specifically to be private.  I suppose we could call it something more specific/obvious but I think we can update that later, as it's a bigger and mostly cosmetic change.

tgamblin commented:
@scheibelp: Ok, I think this is done.  I reworked the tests to tests against the mock configurationfile and an override instead of circularly checking for what we're supposed to be testing.  This actually caught a bug with `colify` (it doesn't always respect `cols=<const>`.  I just replaced the call to colify with some code I think is clearer.

tgamblin commented:
@alalazo: basically I agree with all your points but they're either more work than they're worth (showing merged keys) or they're YAML-related ugliness.  I hope you don't mind if we leave them.
